### PR TITLE
Bug fix: Net fixes x2

### PR DESCRIPTION
**Pull request by subtly**

### CODE REVIEW

File: libethereum/EthereumHost.cpp [hunk 1]
```diff
@@ -218,25 +218,32 @@ void EthereumHost::maintainTransactions()
...
+tuple<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>, list<shared_ptr<Session>>> EthereumHost::randomSelection(unsigned _percent, std::function<bool(EthereumPeer*)> const& _allow)
 {
-	pair<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>> ret;
-	ret.second.reserve(peerSessions().size());
-	for (auto const& j: peerSessions())
+	vector<shared_ptr<EthereumPeer>> chosen;
+	vector<shared_ptr<EthereumPeer>> allowed;
+	list<shared_ptr<Session>> sessions;
+	
+	auto ps = peerSessions();
```
Comment (chfast) [inline libethereum/EthereumHost.cpp]: Avoid copy.

File: libethereum/EthereumHost.cpp [hunk 2]
```diff
@@ -218,25 +218,32 @@ void EthereumHost::maintainTransactions()
...
 	{
-		unsigned n = rand() % ret.second.size();
-		ret.first.push_back(std::move(ret.second[n]));
-		ret.second.erase(ret.second.begin() + n);
+		unsigned n = rand() % allowed.size();
+		chosen.push_back(std::move(allowed[n]));
+		allowed.erase(allowed.begin() + n);
 	}
-	return ret;
+	return make_tuple(chosen, allowed, sessions);
```
Comment (chfast) [inline libethereum/EthereumHost.cpp]: std::move should be used here for all arguments.

File: libp2p/Session.cpp [hunk 3]
```diff
@@ -319,10 +319,14 @@ void Session::send(bytes&& _msg)
 
 void Session::write()
 {
-	const bytes& bytes = m_writeQueue[0];
-	m_io->writeSingleFramePacket(&bytes, m_writeQueue[0]);
+	bytes *out;
```
Comment (chfast) [inline libp2p/Session.cpp]: const disappeared.

File: libethereum/EthereumHost.cpp [hunk 4]
```diff
@@ -218,25 +218,32 @@ void EthereumHost::maintainTransactions()
...
 
-pair<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>> EthereumHost::randomSelection(unsigned _percent, std::function<bool(EthereumPeer*)> const& _allow)
+tuple<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>, list<shared_ptr<Session>>> EthereumHost::randomSelection(unsigned _percent, std::function<bool(EthereumPeer*)> const& _allow)
 {
-	pair<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>> ret;
-	ret.second.reserve(peerSessions().size());
-	for (auto const& j: peerSessions())
+	vector<shared_ptr<EthereumPeer>> chosen;
+	vector<shared_ptr<EthereumPeer>> allowed;
+	list<shared_ptr<Session>> sessions;
```
Comment (chfast) [inline libethereum/EthereumHost.cpp]: Any reason list is used in this case?

File: libp2p/Session.cpp (hunk 3 above)
Comment (subtly) [inline libp2p/Session.cpp]: This can't be const as it's assigned within the Guard.

Comment (chfast) [inline libp2p/Session.cpp]: Pointer itself cannot be (`bytes* const`) but can point a const bytes (`bytes const*`).

File: libethereum/EthereumHost.cpp (hunk 4 above)
Comment (subtly) [inline libethereum/EthereumHost.cpp]: It doesn't need any particular data structure.

Comment (chfast) [inline libethereum/EthereumHost.cpp]: List is almost worse than vector.

File: libethereum/EthereumHost.cpp (hunk 2 above)
Comment (subtly) [inline libethereum/EthereumHost.cpp]: My understanding is that make_tuple moves constructor arguments (&&).

File: libp2p/Session.cpp (hunk 3 above)
Comment (subtly) [inline libp2p/Session.cpp]: Ah. Fixed.

File: libethereum/EthereumHost.cpp (hunk 2 above)
Comment (chfast) [inline libethereum/EthereumHost.cpp]: That's somehow correct. make_tuple moves arguments but only if you pass a rvalue reference. Currently you're passing lvalue references as all arguments have a name. To make them a rvalue reference you need to cast. The cast instruction is just `std::move`.

File: libp2p/Session.cpp (hunk 3 above)
Comment (gavofyork) [inline libp2p/Session.cpp]: whitespace wrong, too.

**Comment by gavofyork** [comment]
all chfast's comments should be addressed.

**Comment by subtly** [comment]
All fixed! buildbots pending.

**Pull request merged**